File: backend/api/routers/conversations.py
from typing import Optional

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile
from sqlalchemy.orm import Session

from backend.core.database import get_db
from backend.api.schemas import SendMessageRequest
from backend.services.entities import conversations, agents
from backend.services.messaging import outbound
from backend.auth.models import AuthUser
from backend.auth.dependencies import get_current_user
from backend.auth import service as auth_service

router = APIRouter(tags=["conversations"])


def require_conversation_access(conv_id: int, user: AuthUser, db: Session):
    """Check conversation access or raise 403."""
    if not auth_service.can_access_conversation(db, user, conv_id):
        raise HTTPException(status_code=403, detail="Access denied to this conversation")


def require_agent_access(agent_id: int, user: AuthUser, db: Session):
    if not auth_service.can_access_agent(db, user, agent_id):
        raise HTTPException(status_code=403, detail="You don't have access to this agent")


def _raise_outbound(err: outbound.OutboundError):
    raise HTTPException(status_code=err.status_code, detail=err.detail)


def _load_conv(db: Session, conv_id: int):
    conv = conversations.get_by_id(db, conv_id)
    if not conv:
        raise HTTPException(status_code=404, detail="Conversation not found")
    return conv


async def _read_upload(upload: Optional[UploadFile]) -> tuple[Optional[bytes], str, str]:
    if not upload or not (upload.filename or "").strip():
        return None, "", ""
    data = await upload.read()
    if not data:
        raise HTTPException(status_code=400, detail="קובץ ריק")
    return data, upload.content_type or "", upload.filename or "file"


@router.get("/inbox/whatsapp")
def whatsapp_inbox(
    agent_id: int = Query(...),
    current_user: AuthUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Active WhatsApp channel plus approved templates for the composer."""
    from backend.models.whatsapp_template import WhatsAppTemplate
    from backend.services.messaging.templates import template_to_dict

    require_agent_access(agent_id, current_user, db)
    agent = agents.get_by_id(db, agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")

    channel = outbound.active_whatsapp_channel(db, agent)
    channel_type = None
    if channel:
        channel_type = channel.channel_type
    elif agent.provider == "wasender":
        channel_type = "whatsapp_wasender"
    elif agent.provider == "meta":
        channel_type = "whatsapp_meta"

    templates = []
    if channel_type == "whatsapp_meta":
        rows = (
            db.query(WhatsAppTemplate)
            .filter(
                WhatsAppTemplate.agent_id == agent_id,
                WhatsAppTemplate.status == "APPROVED",
            )
            .order_by(WhatsAppTemplate.name)
            .all()
        )
        templates = [template_to_dict(t) for t in rows]

    return {"channel_type": channel_type, "templates": templates}


@router.post("/inbox/whatsapp")
async def start_whatsapp_chat(
    agent_id: int = Form(...),
    phone: str = Form(...),
    text: Optional[str] = Form(None),
    caption: Optional[str] = Form(None),
    template_id: Optional[int] = Form(None),
    body_params: Optional[str] = Form(None),
    as_voice: bool = Form(False),
    file: Optional[UploadFile] = File(None),
    header_file: Optional[UploadFile] = File(None),
    current_user: AuthUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Open a WhatsApp chat by phone (existing or new) and send the first message."""
    require_agent_access(agent_id, current_user, db)
    agent = agents.get_by_id(db, agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")

    try:
        conv = await outbound.open_whatsapp_chat(db, agent, phone)
        file_data, file_ct, file_name = await _read_upload(file)
        header_data, header_ct, header_name = await _read_upload(header_file)

        if template_id:
            msg = await outbound.send_template_message(
                db, conv, template_id,
                outbound.parse_body_params(body_params),
                header_data, header_ct, header_name,
            )
        elif file_data:
            msg = await outbound.send_bytes(
                db, conv, file_data, file_ct, file_name, caption, as_voice,
            )
        elif text and text.strip():
            msg = await outbound.send_text(db, conv, text)
        else:
            raise outbound.OutboundError("אין תוכן לשליחה")
    except outbound.OutboundError as e:
        _raise_outbound(e)

    return {"status": "sent", "conversation_id": conv.id, "message_id": msg.id}


@router.get("/{conv_id}/messages")
def list_messages(
    conv_id: int,
    current_user: AuthUser = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get messages for a conversation (must have access)."""
    from backend.services.messaging import messages

    require_conversation_access(conv_id, current_user, db)
    return messages.get_history(db, conv_id)


@router.post("/{conv_id}/send")
async def send_message(
    conv_id: int,
    req: SendMessageRequest,
    current_user: AuthUser = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Send a manual text message. Meta WhatsApp outside the 24h window is rejected."""
    require_conversation_access(conv_id, current_user, db)
    try:
        msg = await outbound.send_text(db, _load_conv(db, conv_id), req.text)
    except outbound.OutboundError as e:
        _raise_outbound(e)
    return {"status": "sent", "message_id": msg.id, "conversation_id": conv_id}


@router.post("/{conv_id}/send-media")
async def send_media(
    conv_id: int,
    caption: Optional[str] = Form(None),
    as_voice: bool = Form(False),
    file: UploadFile = File(...),
    current_user: AuthUser = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Send an image, video, document, or voice note from the browser."""
    require_conversation_access(conv_id, current_user, db)
    data, content_type, filename = await _read_upload(file)
    try:
        msg = await outbound.send_bytes(
            db, _load_conv(db, conv_id), data, content_type, filename, caption, as_voice,
        )
    except outbound.OutboundError as e:
        _raise_outbound(e)
    return {"status": "sent", "message_id": msg.id, "conversation_id": conv_id}


@router.post("/{conv_id}/send-template")
async def send_template(
    conv_id: int,
    template_id: int = Form(...),
    body_params: Optional[str] = Form(None),
    header_file: Optional[UploadFile] = File(None),
    current_user: AuthUser = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Send an approved Meta template. Required outside the 24h customer window."""
    require_conversation_access(conv_id, current_user, db)
    header_data, header_ct, header_name = await _read_upload(header_file)
    try:
        msg = await outbound.send_template_message(
            db,
            _load_conv(db, conv_id),
            template_id,
            outbound.parse_body_params(body_params),
            header_data,
            header_ct,
            header_name,
        )
    except outbound.OutboundError as e:
        _raise_outbound(e)
    return {"status": "sent", "message_id": msg.id, "conversation_id": conv_id}


@router.post("/{conv_id}/pause")
def pause_conversation(
    conv_id: int,
    current_user: AuthUser = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Pause AI responses for this conversation."""
    require_conversation_access(conv_id, current_user, db)
    conv = conversations.set_paused(db, conv_id, True)
    if not conv:
        raise HTTPException(status_code=404, detail="Conversation not found")
    return {"status": "paused", "is_paused": True}


@router.post("/{conv_id}/resume")
def resume_conversation(
    conv_id: int,
    current_user: AuthUser = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Resume AI responses for this conversation."""
    require_conversation_access(conv_id, current_user, db)
    conv = conversations.set_paused(db, conv_id, False)
    if not conv:
        raise HTTPException(status_code=404, detail="Conversation not found")
    return {"status": "resumed", "is_paused": False}


@router.delete("/{conv_id}")
def delete_conversation(
    conv_id: int,
    current_user: AuthUser = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Delete a conversation (must have access)."""
    require_conversation_access(conv_id, current_user, db)
    if not conversations.delete(db, conv_id):
        raise HTTPException(status_code=404, detail="Conversation not found")
    return {"status": "deleted"}

File: frontend/lib/api/conversations.ts
import type { Conversation, Message } from '../types';
import { API_URL, authFetch } from './client';

export interface ConversationCursor {
  cursor_time: string;
  cursor_id: number;
}

export interface ConversationsPage {
  items: Conversation[];
  next_cursor: ConversationCursor | null;
}

export async function getConversations(
  agentId: number,
  cursor?: ConversationCursor | null,
): Promise<ConversationsPage> {
  const params = new URLSearchParams();
  if (cursor) {
    params.set('cursor_time', cursor.cursor_time);
    params.set('cursor_id', String(cursor.cursor_id));
  }
  const qs = params.toString();
  const url = `${API_URL}/api/agents/${agentId}/conversations${qs ? `?${qs}` : ''}`;
  const res = await authFetch(url);
  if (!res.ok) throw new Error('Failed to fetch conversations');
  return res.json();
}

export async function getMessages(convId: number): Promise<Message[]> {
  const res = await authFetch(`${API_URL}/api/conversations/${convId}/messages`);
  if (!res.ok) throw new Error('Failed to fetch messages');
  return res.json();
}

export async function deleteConversation(convId: number): Promise<void> {
  const res = await authFetch(`${API_URL}/api/conversations/${convId}`, { method: 'DELETE' });
  if (!res.ok) throw new Error('Failed to delete conversation');
}

export async function pauseConversation(convId: number): Promise<void> {
  const res = await authFetch(`${API_URL}/api/conversations/${convId}/pause`, { method: 'POST' });
  if (!res.ok) throw new Error('Failed to pause conversation');
}

export async function resumeConversation(convId: number): Promise<void> {
  const res = await authFetch(`${API_URL}/api/conversations/${convId}/resume`, { method: 'POST' });
  if (!res.ok) throw new Error('Failed to resume conversation');
}

async function readApiError(res: Response, fallback: string): Promise<string> {
  const error = await res.json().catch(() => ({ detail: fallback }));
  const detail = error.detail;
  if (typeof detail === 'string') return detail;
  if (Array.isArray(detail)) return detail.map((x: { msg?: string }) => x.msg).filter(Boolean).join(', ') || fallback;
  return fallback;
}

export async function sendMessage(convId: number, text: string): Promise<{ status: string; message_id: number }> {
  const res = await authFetch(`${API_URL}/api/conversations/${convId}/send`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ text }),
  });
  if (!res.ok) throw new Error(await readApiError(res, 'שליחת ההודעה נכשלה'));
  return res.json();
}

export async function sendConversationMedia(
  convId: number,
  file: File,
  caption?: string,
  asVoice?: boolean,
): Promise<{ status: string; message_id: number }> {
  const fd = new FormData();
  fd.append('file', file);
  if (caption) fd.append('caption', caption);
  if (asVoice) fd.append('as_voice', 'true');
  const res = await authFetch(`${API_URL}/api/conversations/${convId}/send-media`, {
    method: 'POST',
    body: fd,
  });
  if (!res.ok) throw new Error(await readApiError(res, 'שליחת המדיה נכשלה'));
  return res.json();
}

export async function sendConversationTemplate(
  convId: number,
  templateId: number,
  bodyParams: string[],
  headerFile?: File,
): Promise<{ status: string; message_id: number }> {
  const fd = new FormData();
  fd.append('template_id', String(templateId));
  fd.append('body_params', JSON.stringify(bodyParams));
  if (headerFile) fd.append('header_file', headerFile);
  const res = await authFetch(`${API_URL}/api/conversations/${convId}/send-template`, {
    method: 'POST',
    body: fd,
  });
  if (!res.ok) throw new Error(await readApiError(res, 'שליחת התבנית נכשלה'));
  return res.json();
}

export interface WhatsAppInbox {
  channel_type: 'whatsapp_wasender' | 'whatsapp_meta' | null;
  templates: import('../types').WhatsAppTemplate[];
}

export function whatsappKindFromAgent(agent: {
  provider?: string | null;
  active_channel_types?: string[] | null;
}): WhatsAppInbox['channel_type'] {
  const types = agent.active_channel_types || [];
  if (types.includes('whatsapp_wasender')) return 'whatsapp_wasender';
  if (types.includes('whatsapp_meta')) return 'whatsapp_meta';
  if (agent.provider === 'wasender') return 'whatsapp_wasender';
  if (agent.provider === 'meta') return 'whatsapp_meta';
  return null;
}

export async function getWhatsAppInbox(agentId: number): Promise<WhatsAppInbox> {
  const res = await authFetch(`${API_URL}/api/conversations/inbox/whatsapp?agent_id=${agentId}`);
  if (!res.ok) throw new Error('Failed to fetch inbox settings');
  return res.json();
}

export async function startWhatsAppChat(input: {
  agentId: number;
  phone: string;
  text?: string;
  file?: File;
  caption?: string;
  asVoice?: boolean;
  templateId?: number;
  bodyParams?: string[];
  headerFile?: File;
}): Promise<{ status: string; conversation_id: number; message_id: number }> {
  const fd = new FormData();
  fd.append('agent_id', String(input.agentId));
  fd.append('phone', input.phone);
  if (input.text) fd.append('text', input.text);
  if (input.caption) fd.append('caption', input.caption);
  if (input.asVoice) fd.append('as_voice', 'true');
  if (input.file) fd.append('file', input.file);
  if (input.templateId) {
    fd.append('template_id', String(input.templateId));
    fd.append('body_params', JSON.stringify(input.bodyParams || []));
  }
  if (input.headerFile) fd.append('header_file', input.headerFile);
  const res = await authFetch(`${API_URL}/api/conversations/inbox/whatsapp`, {
    method: 'POST',
    body: fd,
  });
  if (!res.ok) throw new Error(await readApiError(res, 'פתיחת הצ׳אט נכשלה'));
  return res.json();
}
